File: grind75/Jan13/105.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
    if len(preorder) == 0 { return nil }
    val := preorder[0]
    root := &TreeNode{Val: val}
    idx := findIdx(inorder, val)
    root.Left = buildTree(preorder[1:idx+1], inorder[:idx])
    root.Right = buildTree(preorder[idx+1:], inorder[idx+1:])
    return root
}

func findIdx(arr []int, val int) int {
    for i:= 0 ; i < len(arr); i++ {
        if arr[i] == val {
            return i
        }
    }
    return -1
}

File: grind75/Jan13/105.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> 
Optional[TreeNode]:
        if inorder:
            root = preorder.pop(0)
            root_index = inorder.index(root)
            left = inorder[:root_index]
            right = inorder[root_index+1:]
            root_node = TreeNode(root)
            root_node.left = self.buildTree(preorder, left)
            root_node.right = self.buildTree(preorder, right)
            return root_node
